### ISSUE TITLE
Is it possible to document OAuth 2 scopes for different HTTP methods?

### ISSUE THREAD

**Issue by Beontra** [springfox/springfox#2180]
We already managed to document the scope for different paths.
However, we would like to use different scopes for different HTTP methods, that is, for GET/POST/PUT/DELETE.

So far we only managed to select paths within the SecurityContext builder:
`SecurityContext.builder()
            .securityReferences(defaultAuth())
            .forPaths(PathSelectors.ant("/api/*))
            .build())`

Is it also possible to add the HTTP method there?

**Comment by dilipkrish** [comment]
@Beontra thanks for the feature request

**Comment by imre-kerr** [comment]
This is implemented now, with the `forHttpMethods` method:
```kotlin
SecurityContext.builder()
    .forHttpMethods { t -> t?.matches("POST") ?: false }
    .forPaths(PathSelectors.regex("/api/.*"))
    .securityReferences(listOf(SecurityReference("Auth0", arrayOf(writeAll))))
    .build()
```
However it fails for me out of the box, with an NPE:
```
2018-06-04 09:08:56.120 ERROR [-,,,] 12701 --- [           main] d.s.w.p.DocumentationPluginsBootstrapper : Unable to scan documentation context default

com.google.common.util.concurrent.UncheckedExecutionException: java.lang.NullPointerException
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2214) ~[guava-20.0.jar:na]
	at com.google.common.cache.LocalCache.get(LocalCache.java:4053) ~[guava-20.0.jar:na]
	at com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:4057) ~[guava-20.0.jar:na]
	at com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4986) ~[guava-20.0.jar:na]
	at com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4992) ~[guava-20.0.jar:na]
	at springfox.documentation.spring.web.scanners.CachingOperationReader.read(CachingOperationReader.java:57) ~[springfox-spring-web-2.9.0.jar:2.9.0]
	at springfox.documentation.spring.web.scanners.ApiDescriptionReader.read(ApiDescriptionReader.java:70) ~[springfox-spring-web-2.9.0.jar:2.9.0]
	at springfox.documentation.spring.web.scanners.ApiListingScanner.scan(ApiListingScanner.java:134) ~[springfox-spring-web-2.9.0.jar:2.9.0]
	at springfox.documentation.spring.web.scanners.ApiDocumentationScanner.scan(ApiDocumentationScanner.java:71) ~[springfox-spring-web-2.9.0.jar:2.9.0]
	at springfox.documentation.spring.web.plugins.DocumentationPluginsBootstrapper.scanDocumentation(DocumentationPluginsBootstrapper.java:101) [springfox-spring-web-2.9.0.jar:2.9.0]
	at springfox.documentation.spring.web.plugins.DocumentationPluginsBootstrapper.start(DocumentationPluginsBootstrapper.java:167) [springfox-spring-web-2.9.0.jar:2.9.0]
	at org.springframework.context.support.DefaultLifecycleProcessor.doStart(DefaultLifecycleProcessor.java:181) [spring-context-5.0.6.RELEASE.jar:5.0.6.RELEASE]
	at org.springframework.context.support.DefaultLifecycleProcessor.access$200(DefaultLifecycleProcessor.java:52) [spring-context-5.0.6.RELEASE.jar:5.0.6.RELEASE]
	at org.springframework.context.support.DefaultLifecycleProcessor$LifecycleGroup.start(DefaultLifecycleProcessor.java:356) [spring-context-5.0.6.RELEASE.jar:5.0.6.RELEASE]
	at org.springframework.context.support.DefaultLifecycleProcessor.startBeans(DefaultLifecycleProcessor.java:157) [spring-context-5.0.6.RELEASE.jar:5.0.6.RELEASE]
	at org.springframework.context.support.DefaultLifecycleProcessor.onRefresh(DefaultLifecycleProcessor.java:121) [spring-context-5.0.6.RELEASE.jar:5.0.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:885) [spring-context-5.0.6.RELEASE.jar:5.0.6.RELEASE]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.finishRefresh(ServletWebServerApplicationContext.java:161) [spring-boot-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:553) [spring-context-5.0.6.RELEASE.jar:5.0.6.RELEASE]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:140) [spring-boot-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:759) [spring-boot-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:395) [spring-boot-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:327) [spring-boot-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1255) [spring-boot-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1243) [spring-boot-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at no.nsb.prm.pricecalendar.PriceCalendarApplicationKt.main(PriceCalendarApplication.kt:13) [classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_171]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_171]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_171]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_171]
	at org.springframework.boot.maven.AbstractRunMojo$LaunchRunner.run(AbstractRunMojo.java:496) [spring-boot-maven-plugin-2.0.2.RELEASE.jar:2.0.2.RELEASE]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_171]
Caused by: java.lang.NullPointerException: null
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:770) ~[guava-20.0.jar:na]
	at com.google.common.collect.FluentIterable.<init>(FluentIterable.java:125) ~[guava-20.0.jar:na]
	at com.google.common.collect.FluentIterable$1.<init>(FluentIterable.java:143) ~[guava-20.0.jar:na]
	at com.google.common.collect.FluentIterable.from(FluentIterable.java:141) ~[guava-20.0.jar:na]
	at springfox.documentation.swagger.readers.operation.OperationAuthReader.apply(OperationAuthReader.java:63) ~[springfox-swagger-common-2.9.0.jar:2.9.0]
	at springfox.documentation.spring.web.plugins.DocumentationPluginsManager.operation(DocumentationPluginsManager.java:120) ~[springfox-spring-web-2.9.0.jar:2.9.0]
	at springfox.documentation.spring.web.readers.operation.ApiOperationReader.read(ApiOperationReader.java:73) ~[springfox-spring-web-2.9.0.jar:2.9.0]
	at springfox.documentation.spring.web.scanners.CachingOperationReader$1.load(CachingOperationReader.java:50) ~[springfox-spring-web-2.9.0.jar:2.9.0]
	at springfox.documentation.spring.web.scanners.CachingOperationReader$1.load(CachingOperationReader.java:48) ~[springfox-spring-web-2.9.0.jar:2.9.0]
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3628) ~[guava-20.0.jar:na]
	at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2336) ~[guava-20.0.jar:na]
	at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2295) ~[guava-20.0.jar:na]
	at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2208) ~[guava-20.0.jar:na]
	... 31 common frames omitted
```
The fix is pretty simple: https://github.com/imre-kerr/springfox/commit/ecaa24bdce632b6dd431a2cc282329a3289bb2f8

**Comment by dilipkrish** [comment]
Thanks for catching that @imre-kerr

**Comment by dilipkrish** [comment]
Fixed via #2502
